codec/src/main/java/io/netty/handler/codec/BulkLengthFieldBasedFrameDecoder.java

```diff
@@ -0,0 +1,160 @@
+     * {@link TooLongFrameException} will be thrown.
+     * @param lengthFieldOffset the offset of the length field
+     * @param lengthFieldLength the length of the length field
+     */
+    public BulkLengthFieldBasedFrameDecoder(int startBufferSize, int maxFrameLength, int lengthFieldOffset,
+                                            int lengthFieldLength) {
+        super(startBufferSize);
+
+        if (maxFrameLength <= 0) {
+            throw new IllegalArgumentException(
+                    "maxFrameLength must be a positive integer: " +
+                    maxFrameLength);
+        }
+
+        if (lengthFieldOffset < 0) {
+            throw new IllegalArgumentException(
+                    "lengthFieldOffset must be a non-negative integer: " +
+                    lengthFieldOffset);
+        }
+
+        if (lengthFieldOffset > maxFrameLength - lengthFieldLength) {
+            throw new IllegalArgumentException(
+                    "maxFrameLength (" + maxFrameLength + ") " +
+                    "must be equal to or greater than " +
+                    "lengthFieldOffset (" + lengthFieldOffset + ") + " +
+                    "lengthFieldLength (" + lengthFieldLength + ").");
+        }
+
+        this.maxFrameLength = maxFrameLength;
+        this.lengthFieldOffset = lengthFieldOffset;
+        this.lengthFieldLength = lengthFieldLength;
+        lengthFieldEndOffset = lengthFieldOffset + lengthFieldLength;
+    }
+
+    @Override
+    protected void decode(ChannelHandlerContext ctx, ByteBuf buffer, List<Object> out) {
+        int readableBytes = buffer.readableBytes();
+        int rdIdx = buffer.readerIndex();
+
+        while (readableBytes >= lengthFieldEndOffset) {
+            int actualLengthFieldOffset = rdIdx + lengthFieldOffset;
+            long frameLength = getUnadjustedFrameLength(buffer, actualLengthFieldOffset, lengthFieldLength);
+
+            if (frameLength < 0) {
+                throw new CorruptedFrameException(
+                        "negative pre-adjustment length field: " + frameLength);
+            }
+
+            frameLength += lengthFieldEndOffset;
+
+            if (frameLength > maxFrameLength) {
+                throw new TooLongFrameException(
+                        "Adjusted frame length exceeds " + maxFrameLength +
+                        ": " + frameLength + " - discarded");
+            }
+
+            int frameLengthInt = (int) frameLength;
+
+            if (readableBytes < frameLengthInt) {
+                break;
+            }
+
+            readableBytes -= frameLengthInt;
+            rdIdx += frameLengthInt;
+        }
+
+        if (rdIdx == 0) {
+            return;
+        }
+
+        final int previousWriterIndex = buffer.writerIndex();
+
+        buffer.writerIndex(rdIdx);
+
+        final ByteBuf chunk = ctx.alloc().buffer(rdIdx, rdIdx).writeBytes(buffer, rdIdx);
```

Comment:
Do you mean `buffer.retain().slice()`? Well, it's not how it supposed to be working. The whole point of this decoder is to batch all incoming ByteBuf messages into one preallocated ByteBuf object. When the `channelReadComplete` method gets called we simply copy data from buffer to another chunk buffer which would contain precisely N messages (where N >= 1) and pass it to next pipeline element. By doing so we're able to reuse this buffer again and again no matter how many messages we get. If I chose retain-slice approach I would have to keep an eye on the buffer reference counter, because slice creates only a view of buffer sub-region, so you wouldn't be able to reuse it, if for example, the channel wasn't ready for write (channel.isWritable() returns false).